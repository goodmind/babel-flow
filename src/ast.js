// @flow

import execa from "execa";
import fs from "fs-extra";
import * as parser from "@babel/parser";
import traverse from "@babel/traverse";
import template from "@babel/template";

function flowBinVersion(): string | null {
  try {
    //$FlowFixMe
    const version = require("flow-bin/package.json").version;
    return version;
  } catch (e) {
    return null;
  }
}

const flowVersion = flowBinVersion();

async function getTypeAtPos(
  { line = 0, column = 0 } = { line: 0, column: 0 },
  filename: string
) {
  const command = `npx flow type-at-pos --expand-type-aliases --json ${filename} ${line} ${column +
    1}`;
  const result = await execa.shell(command);
  return JSON.parse(result.stdout);
}

function getTypeAtPosSync(
  { line = 0, column = 0 } = { line: 0, column: 0 },
  filename: string
) {
  const command = `npx flow type-at-pos --expand-type-aliases --json ${filename} ${line} ${column +
    1}`;
  return JSON.parse(execa.shellSync(command).stdout);
}

function getTypeAtPosExpandedSync(
  { line = 0, column = 0 } = { line: 0, column: 0 },
  filename: string
) {
  const command = `npx flow type-at-pos --expand-type-aliases --expand-json-output ${filename} ${line} ${column +
    1}`;
  return JSON.parse(execa.shellSync(command).stdout);
}

const plugins = ["flow", "flowComments"];

function convertType(type: string, plugins: string[]): string {
  if (type === "any (explicit)") return "any";
  if (type === "any (implicit)") return "any";
  if (type.startsWith("class ")) {
    const ast = template.ast(`${type} {}`, {
      plugins
    });
    return `typeof ${ast.id.name}`;
  }
  return type;
}

export function parseFileSync(
  filename: string,
  {
    sourceType = "module",
    plugins = plugins,
    ...options
  }: { [key: string]: any } = { sourceType: "module", plugins }
) {
  const code = fs.readFileSync(filename, { encoding: "utf8" });
  const ast = parser.parse(code, {
    ...options,
    plugins,
    sourceType
  });

  if (flowVersion === null) {
    throw new Error("flow-bin must be installed");
  }

  function assignType(path) {
    const symbol = getTypeAtPosSync(path.node.loc.start, filename);
    const type = convertType(symbol.type, plugins);
    //console.log(symbol.type);
    if (type !== "(unknown)") {
      try {
        symbol.type = template.ast(`type S = ${type}`, {
          plugins
        }).right;
        path.node.symbol = symbol;
      } catch (err) {
        console.error("Error on ", path.node.loc.start, path.node);
        console.error(err);
      }
    }
  }

  traverse(ast, {
    enter(path) {
      assignType(path);
    }
  });

  return ast;
}

export async function parseFile(
  filename: string,
  {
    sourceType = "module",
    plugins = plugins,
    ...options
  }: { [key: string]: any } = { sourceType: "module", plugins }
) {
  const code = await fs.readFile(filename, { encoding: "utf8" });
  const ast = parser.parse(code, {
    ...options,
    sourceType,
    plugins
  });

  if (flowVersion === null) {
    throw new Error("flow-bin must be installed");
  }

  let promises = [];

  async function assignType(path) {
    const symbol = await getTypeAtPos(path.node.loc.start, filename);
    const type = convertType(symbol.type, plugins);
    //console.log(symbol);
    if (type !== "(unknown)") {
      try {
        symbol.type = template.ast(`type S = ${type}`, {
          plugins
        }).right;
        path.node.symbol = symbol;
      } catch (err) {
        console.error("Error on ", path.node.loc.start, path.node);
        console.error(err);
      }
    }
  }

  traverse(ast, {
    enter(path) {
      promises.push(assignType(path));
    }
  });

  await Promise.all(promises);

  return ast;
}
