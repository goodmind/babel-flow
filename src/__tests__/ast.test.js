// @flow

import { parseFile } from "../ast";
import { benchAsync } from "./bench";
import generate from "@babel/generator";
import path from "path";

const fixture = path.join(__dirname, "./magic.js");

test("should parse file", async () => {
  const ast = await parseFile(fixture);
  expect(ast).toMatchSnapshot();
  expect(generate(ast).code).toMatchSnapshot();
}, 100000);

// test("time", async () => {
//   expect(await benchAsync(() => parseFile(fixture))).toMatchSnapshot();
// }, 100000);
