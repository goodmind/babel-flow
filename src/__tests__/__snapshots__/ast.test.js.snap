// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse file 1`] = `
Node {
  "comments": Array [
    Object {
      "end": 7,
      "loc": SourceLocation {
        "end": Position {
          "column": 7,
          "line": 1,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "start": 0,
      "type": "CommentLine",
      "value": "@flow",
    },
  ],
  "end": 35,
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 4,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "declaration": Node {
          "declarations": Array [
            Node {
              "end": 33,
              "id": Node {
                "end": 27,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 18,
                    "line": 3,
                  },
                  "identifierName": "magic",
                  "start": Position {
                    "column": 13,
                    "line": 3,
                  },
                },
                "name": "magic",
                "start": 22,
                "type": "Identifier",
              },
              "init": Node {
                "end": 33,
                "extra": Object {
                  "raw": "\\"a\\"",
                  "rawValue": "a",
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 24,
                    "line": 3,
                  },
                  "start": Position {
                    "column": 21,
                    "line": 3,
                  },
                },
                "start": 30,
                "type": "StringLiteral",
                "value": "a",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 24,
                  "line": 3,
                },
                "start": Position {
                  "column": 13,
                  "line": 3,
                },
              },
              "start": 22,
              "symbol": Object {
                "end": 18,
                "endline": 3,
                "line": 3,
                "loc": Object {
                  "end": Object {
                    "column": 18,
                    "line": 3,
                    "offset": 27,
                  },
                  "start": Object {
                    "column": 14,
                    "line": 3,
                    "offset": 22,
                  },
                  "type": "SourceFile",
                },
                "reasons": Array [],
                "start": 14,
                "type": Object {
                  "loc": undefined,
                  "type": "StringTypeAnnotation",
                },
              },
              "type": "VariableDeclarator",
            },
          ],
          "end": 34,
          "kind": "const",
          "loc": SourceLocation {
            "end": Position {
              "column": 25,
              "line": 3,
            },
            "start": Position {
              "column": 7,
              "line": 3,
            },
          },
          "start": 16,
          "type": "VariableDeclaration",
        },
        "end": 34,
        "exportKind": "value",
        "leadingComments": Array [
          Object {
            "end": 7,
            "loc": SourceLocation {
              "end": Position {
                "column": 7,
                "line": 1,
              },
              "start": Position {
                "column": 0,
                "line": 1,
              },
            },
            "start": 0,
            "type": "CommentLine",
            "value": "@flow",
          },
        ],
        "loc": SourceLocation {
          "end": Position {
            "column": 25,
            "line": 3,
          },
          "start": Position {
            "column": 0,
            "line": 3,
          },
        },
        "source": null,
        "specifiers": Array [],
        "start": 9,
        "type": "ExportNamedDeclaration",
      },
    ],
    "directives": Array [],
    "end": 35,
    "interpreter": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 0,
        "line": 4,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  "start": 0,
  "type": "File",
}
`;

exports[`should parse file 2`] = `
"//@flow
export const magic = \\"a\\";"
`;
